File: task-1/VertexCover.java
/*
 * THIS VERSION IS NOT OPTIMISED. IT WAS PRORAMMED AS "BADLY" AS POSSIBLE ...
 * NOW WE HAVE TO THINK ABOUT SOME OPTIMIZATION STRATEGIES !!!
 */

import java.util.*;

public class VertexCover {
    public static ArrayList<String> vertex_cover(ArrayList<String> nodes,ArrayList<String[]> edges){
        // Test for vertex-cover of size k=[0 ... |V|]
        for (int k=0;k<nodes.size();k++){
            ArrayList<String> S = vc_branch(nodes,edges,k);
            if (S != null) return S;
        }
        return nodes;
    }
    public static ArrayList<String> vc_branch(ArrayList<String> nodes,ArrayList<String[]> edges,int k){
        if (k < 0) return null;
        else if (edges.size() == 0) return new ArrayList<>();
        for (String node : edges.get(0)){
            // Copy nodes list (but remove current 'node')
            ArrayList<String> nodes2 = new ArrayList<>();
            for (String n : nodes) {
                if (!n.equals(node)) nodes2.add(n);
            }
            // Copy edges list (but remove edges containing current 'node')
            ArrayList<String[]> edges2 = new ArrayList<>();
            for (String[] edge : edges){
                if (!(edge[0].equals(node) || edge[1].equals(node))) edges2.add(edge);
            }
            // Test reduced graph for vertex-cover (size k-1)
            ArrayList<String> S = vc_branch(nodes2, edges2, k-1);
            if (S != null) {
                S.add(node);
                return S;
            }
        }
        return null;
    }
    public static void main(String[] args){
        // STEP 1: Parse Input
        Scanner scanner = new Scanner(System.in);
        int nodes_num = Integer.parseInt(scanner.nextLine().split("\\s")[1]);
        int edges_num = Integer.parseInt(scanner.nextLine().split("\\s")[1]);
        ArrayList<String[]> edges = new ArrayList<>();                  // edges => ArrayList of 2-dimensional String-Array
        ArrayList<String> nodes = new ArrayList<>();                    // nodes => 1-dimensional String-Array
        for (int i=0;i<edges_num;i++){
            String[] edge = scanner.nextLine().split("\\s");
            edges.add(edge);
            if (!nodes.contains(edge[0])) nodes.add(edge[0]);
            if (!nodes.contains(edge[1])) nodes.add(edge[1]);
        }
        scanner.close();
        // STEP 2: Run vertex-cover algorithm
        ArrayList<String> result = vertex_cover(nodes,edges);
        // Step 3: Print result
        for (String node : result){
            System.out.println(node);
        }
    }
}

File: task-1/vertex-cover.py
import re

DEBUG = True			# True: Enable comments, False: disable comments

# Main method
def main():
	V,E = read_input()	# Parse input from stdin
	S = vertex_cover(V,E)	# Execute V.C. algorithm
	print("\n".join(S))	# Print result

# Parse input graph from stdin
def read_input():
	nodes = int(re.match(r"# (\d+) vertices",input()).group(1))	# Parse number of nodes
	edges = int(re.match(r"# (\d+) edges",input()).group(1))	# Parse number of edges
	V,E = set(),list()
	for i in range(edges):						# Parse each following edge line
		n1,n2 = input().split()
		E.append((n1,n2))
		V.add(n1)
		V.add(n2)
	return V,E

# Find minimal vertex cover method (from lecture)
def vertex_cover(V,E):
	for k in range(0,len(V)):
		if DEBUG: print("# k =",k)
		S = vc_branch(V,E,k)
		if S != None:
			return S
	return V

# Find vertex cover (size k) method (from lecture)
def vc_branch(V,E,k):
	if DEBUG: print("# V={V}  E={E}  k={k}".format(V=V,E=E,k=k))
	if k < 0: return None
	elif len(E) == 0: return set()
	for n in E[0]:
		Vn,En = rem(V,E,n)
		S = vc_branch(Vn,En,k-1)
		if S != None:
			S.add(n)
			return S
	return None

# Remove a node n from graph G=(V,E)
def rem(V,E,v):
	V2 = {n for n in V if n != v}
	E2 = [e for e in E if v not in e]
	return V2,E2

# Call main method
main()
